File: java/Samples/Biometrics/VB/AbisSampleVB/Settings/SettingsManager.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Devices
Imports Neurotec.Samples.My

Public Enum ConnectionType
	SQLiteDatabase
	OdbcDatabase
	RemoteMatchingServer
End Enum

Public Class SampleDbSchema
#Region "Public constructor"

	Public Sub New()
		BiographicData = New NBiographicDataSchema()
		CustomData = New NBiographicDataSchema()
	End Sub

#End Region

#Region "Public properties"

	Private privateBiographicData As NBiographicDataSchema
	Public Property BiographicData() As NBiographicDataSchema
		Get
			Return privateBiographicData
		End Get
		Set(ByVal value As NBiographicDataSchema)
			privateBiographicData = value
		End Set
	End Property
	Private privateCustomData As NBiographicDataSchema
	Public Property CustomData() As NBiographicDataSchema
		Get
			Return privateCustomData
		End Get
		Set(ByVal value As NBiographicDataSchema)
			privateCustomData = value
		End Set
	End Property
	Private privateGenderDataName As String
	Public Property GenderDataName() As String
		Get
			Return privateGenderDataName
		End Get
		Set(ByVal value As String)
			privateGenderDataName = value
		End Set
	End Property
	Private privateEnrollDataName As String
	Public Property EnrollDataName() As String
		Get
			Return privateEnrollDataName
		End Get
		Set(ByVal value As String)
			privateEnrollDataName = value
		End Set
	End Property
	Private privateThumbnailDataName As String
	Public Property ThumbnailDataName() As String
		Get
			Return privateThumbnailDataName
		End Get
		Set(ByVal value As String)
			privateThumbnailDataName = value
		End Set
	End Property
	Private privateSchemaName As String
	Public Property SchemaName() As String
		Get
			Return privateSchemaName
		End Get
		Set(ByVal value As String)
			privateSchemaName = value
		End Set
	End Property
	Public ReadOnly Property HasCustomData() As Boolean
		Get
			Return CustomData IsNot Nothing AndAlso CustomData.Elements.Count > 0
		End Get
	End Property
	Public ReadOnly Property IsEmpty() As Boolean
		Get
			Return Me Is Empty
		End Get
	End Property

#End Region

#Region "Public fields"

	Public Shared ReadOnly Empty As New SampleDbSchema() With {.SchemaName = "None"}

#End Region

#Region "Public methods"

	Public Overrides Function ToString() As String
		Return SchemaName
	End Function

	Public Shared Function Parse(ByVal value As String) As SampleDbSchema
		Dim sc As New SampleDbSchema()
		Dim values = value.Split(New String() {"#"}, StringSplitOptions.None)
		If values.Length <> 6 Then
			Throw New ArgumentException()
		End If

		sc.SchemaName = values(0)
		sc.BiographicData = NBiographicDataSchema.Parse(values(1))
		If (Not String.IsNullOrEmpty(values(2))) Then
			sc.CustomData = NBiographicDataSchema.Parse(values(2))
		End If

		sc.GenderDataName = values(3).Split(New Char() {"="c})(1)
		sc.ThumbnailDataName = values(4).Split(New Char() {"="c})(1)
		sc.EnrollDataName = values(5).Split(New Char() {"="c})(1)

		Return sc
	End Function

	Public Function Save() As String
		If IsEmpty Then
			Throw New InvalidOperationException()
		End If

		Dim format As String = "{0}#{1}#{2}#Gender={3}#Thumbnail={4}#EnrollData={5}"
		Return String.Format(format, SchemaName, If(CObj(BiographicData), String.Empty), If(CObj(CustomData), String.Empty), GenderDataName, ThumbnailDataName, EnrollDataName)
	End Function

#End Region
End Class

Public NotInheritable Class SettingsManager
#Region "Public methods"

	Private Sub New()
	End Sub
	Public Shared Sub LoadSettings(ByVal client As NBiometricClient)
		Dim s As Settings = Settings.Default
		Dim propertyBag As NPropertyBag = Nothing
		Dim propertiesString As String = String.Empty

		If client Is Nothing Then
			Throw New ArgumentNullException("client")
		End If

		client.Reset()
		client.UseDeviceManager = True
		Try
			propertiesString = My.Settings.ClientProperties
		Catch
		End Try
		propertyBag = NPropertyBag.Parse(propertiesString)
		propertyBag.ApplyTo(client)

		client.FingersDeterminePatternClass = client.FingersDeterminePatternClass AndAlso LicensingTools.CanDetectFingerSegments(client.LocalOperations)
		client.FingersCalculateNfiq = client.FingersCalculateNfiq AndAlso LicensingTools.CanAssessFingerQuality(client.LocalOperations)
		Dim remoteConnection = client.RemoteConnections.FirstOrDefault()
		Dim remoteOperations As NBiometricOperations = If(remoteConnection IsNot Nothing, remoteConnection.Operations, NBiometricOperations.None)
		client.FingersCheckForDuplicatesWhenCapturing = client.FingersCheckForDuplicatesWhenCapturing AndAlso LicensingTools.CanFingerBeMatched(remoteOperations)
		If Not LicensingTools.CanDetectFaceSegments(client.LocalOperations) Then
			client.FacesDetectAllFeaturePoints = False
			client.FacesDetectBaseFeaturePoints = False
			client.FacesDetermineGender = False
			client.FacesRecognizeEmotion = False
			client.FacesDetectProperties = False
			client.FacesRecognizeEmotion = False
			client.FacesDetermineAge = False
		End If
	End Sub

	Public Shared Sub LoadPreferedDevices(ByVal client As NBiometricClient)
		Dim s As Settings = Settings.Default
		Try
			If (Not String.IsNullOrEmpty(s.FingerScanner)) Then
				Dim device As NDevice = client.DeviceManager.Devices.FirstOrDefault(Function(x) x.Id = s.FingerScanner)
				If device IsNot Nothing Then
					client.FingerScanner = CType(device, NFScanner)
				End If
			End If
		Catch
		End Try
		Try
			If (Not String.IsNullOrEmpty(s.PalmScanner)) Then
				Dim device As NDevice = client.DeviceManager.Devices.FirstOrDefault(Function(x) x.Id = s.PalmScanner)
				If device IsNot Nothing Then
					client.PalmScanner = CType(device, NFScanner)
				End If
			End If
		Catch
		End Try
		Try
			If (Not String.IsNullOrEmpty(s.IrisScanner)) Then
				Dim device As NDevice = client.DeviceManager.Devices.FirstOrDefault(Function(x) x.Id = s.IrisScanner)
				If device IsNot Nothing Then
					client.IrisScanner = CType(device, NIrisScanner)
				End If
			End If
		Catch
		End Try
		Try
			If (Not String.IsNullOrEmpty(s.FaceCaptureDevice)) Then
				Dim device As NDevice = client.DeviceManager.Devices.FirstOrDefault(Function(x) x.Id = s.FaceCaptureDevice)
				If device IsNot Nothing Then
					client.FaceCaptureDevice = CType(device, NCamera)
				End If
			End If
		Catch
		End Try
		Try
			If (Not String.IsNullOrEmpty(s.VoiceCaptureDevice)) Then
				Dim device As NDevice = client.DeviceManager.Devices.FirstOrDefault(Function(x) x.Id = s.VoiceCaptureDevice)
				If device IsNot Nothing Then
					client.VoiceCaptureDevice = CType(device, NMicrophone)
				End If
			End If
		Catch
		End Try
	End Sub

	Public Shared Sub SaveSettings(ByVal client As NBiometricClient)
		Dim s As Settings = Settings.Default
		Dim properties As New NPropertyBag()

		If client Is Nothing Then
			Throw New ArgumentNullException("client")
		End If

		client.CaptureProperties(properties)
		s.ClientProperties = properties.ToString()

		' prefered devices
		s.FaceCaptureDevice = If(client.FaceCaptureDevice IsNot Nothing, client.FaceCaptureDevice.Id, Nothing)
		s.FingerScanner = If(client.FingerScanner IsNot Nothing, client.FingerScanner.Id, Nothing)
		s.PalmScanner = If(client.PalmScanner IsNot Nothing, client.PalmScanner.Id, Nothing)
		s.IrisScanner = If(client.IrisScanner IsNot Nothing, client.IrisScanner.Id, Nothing)
		s.VoiceCaptureDevice = If(client.VoiceCaptureDevice IsNot Nothing, client.VoiceCaptureDevice.Id, Nothing)

		s.Save()
	End Sub

#End Region

#Region "Public properties"

	Public Shared Property Phrases() As IEnumerable(Of Phrase)
		Get
			Dim values As String = Nothing
			Try
				values = My.Settings.Phrases
			Catch
			End Try

			Dim result As List(Of Phrase) = New List(Of Phrase)
			If (Not String.IsNullOrEmpty(values)) Then
				Dim split() As String = values.Split(New Char() {";"c}, StringSplitOptions.RemoveEmptyEntries)
				For Each item In split
					Dim phrase As Phrase = Nothing
					Try
						Dim splitPhrase = item.Split("="c)
						phrase = New Phrase(Integer.Parse(splitPhrase(0)), splitPhrase(1))
					Catch
						'ignore invalid entries
						Continue For
					End Try
					result.Add(phrase)
				Next item
			End If
			Return result
		End Get
		Set(ByVal value As IEnumerable(Of Phrase))
			If value IsNot Nothing Then
				Dim sb As New StringBuilder()
				For Each phrase As Phrase In value
					sb.AppendFormat("{0}={1};", phrase.Id, phrase.String)
				Next phrase
				My.Settings.Phrases = sb.ToString()
			Else
				My.Settings.Phrases = String.Empty
			End If
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property ConnectionType() As ConnectionType
		Get
			Try
				Return My.Settings.ConnectionType
			Catch
				Return True
			End Try
		End Get
		Set(ByVal value As ConnectionType)
			My.Settings.ConnectionType = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property OdbcConnectionString() As String
		Get
			Try
				Return My.Settings.OdbcConnectionString
			Catch
				Return String.Empty
			End Try
		End Get
		Set(ByVal value As String)
			My.Settings.OdbcConnectionString = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property TableName() As String
		Get
			Try
				Return My.Settings.TableName
			Catch ex As Exception
				Return String.Empty
			End Try
		End Get
		Set(ByVal value As String)

		End Set
	End Property

	Public Shared Property RemoteServerHostName() As String
		Get
			Try
				Return My.Settings.HostName
			Catch
				Return "localhost"
			End Try
		End Get
		Set(ByVal value As String)
			My.Settings.HostName = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property RemoteServerPort() As Integer
		Get
			Try
				Return My.Settings.ClientPort
			Catch
				Return 25452
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.ClientPort = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property RemoteServerAdminPort() As Integer
		Get
			Try
				Return My.Settings.AdminPort
			Catch
				Return 24932
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.AdminPort = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property FingersGeneralizationRecordCount() As Integer
		Get
			Try
				Return My.Settings.FingersGeneralizationRecordCount
			Catch
				Return 3
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.FingersGeneralizationRecordCount = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property PalmsGeneralizationRecordCount() As Integer
		Get
			Try
				Return My.Settings.PalmsGeneralizationRecordCount
			Catch
				Return 3
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.PalmsGeneralizationRecordCount = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property FacesGeneralizationRecordCount() As Integer
		Get
			Try
				Return My.Settings.FacesGeneralizationRecordCount
			Catch
				Return 3
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.FacesGeneralizationRecordCount = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property QuerySuggestions() As String()
		Get
			Try
				Return My.Settings.QueryAutoComplete.OfType(Of String)().ToArray()
			Catch
				Return New String() {}
			End Try
		End Get
		Set(ByVal value As String())
			My.Settings.QueryAutoComplete = New System.Collections.Specialized.StringCollection()
			If value IsNot Nothing Then
				My.Settings.QueryAutoComplete.AddRange(value)
			End If
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property WarnHasSchema() As Boolean
		Get
			Try
				Return My.Settings.WarnHasSchema
			Catch
				Return True
			End Try
		End Get
		Set(ByVal value As Boolean)
			My.Settings.WarnHasSchema = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property CurrentSchemaIndex() As Integer
		Get
			Try
				Return My.Settings.CurrentScema
			Catch
				Return 0
			End Try
		End Get
		Set(ByVal value As Integer)
			My.Settings.CurrentScema = value
			My.Settings.Save()
		End Set
	End Property

	Public Shared Property Schemas() As IEnumerable(Of SampleDbSchema)
		Get
			Dim schemasLocal() As String = Nothing
			Try
				schemasLocal = My.Settings.Schemas.OfType(Of String)().ToArray()
			Catch
				schemasLocal = New String() {}
			End Try

			Return schemasLocal.Select(Function(x) SampleDbSchema.Parse(x)).ToArray()
		End Get
		Set(ByVal value As IEnumerable(Of SampleDbSchema))
			My.Settings.Schemas.Clear()
			If value IsNot Nothing Then
				For Each item In value
					My.Settings.Schemas.Add(item.Save())
				Next item
			End If
			My.Settings.Save()
		End Set
	End Property

	Public Shared ReadOnly Property CurrentSchema() As SampleDbSchema
		Get
			If CurrentSchemaIndex = -1 Then
				Return SampleDbSchema.Empty
			Else
				Return Schemas.ToArray()(CurrentSchemaIndex)
			End If
		End Get
	End Property

	Public Shared Property LocalOperationsIndex() As Integer
		Get
			Try
				Return Settings.Default.LocalOperations
			Catch
				Return 5
			End Try
		End Get
		Set(ByVal value As Integer)
			Settings.Default.LocalOperations = value
			Settings.Default.Save()
		End Set
	End Property

	Public Shared Property FacesMirrorHorizontally() As Boolean
		Get
			Try
				Return Settings.Default.FacesMirrorHorizontally
			Catch
				Return True
			End Try
		End Get
		Set(ByVal value As Boolean)
			Settings.Default.FacesMirrorHorizontally = value
			Settings.Default.Save()
		End Set
	End Property

#End Region
End Class

File: java/Samples/Biometrics/CS/AbisSampleCS/Settings/SettingsManager.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Devices;
using Neurotec.Samples.Properties;

namespace Neurotec.Samples
{
	public enum ConnectionType
	{
		SQLiteDatabase,
		OdbcDatabase,
		RemoteMatchingServer,
	};

	public class SampleDbSchema
	{
		#region Public constructor

		public SampleDbSchema()
		{
			BiographicData = new NBiographicDataSchema();
			CustomData = new NBiographicDataSchema();
		}

		#endregion

		#region Public properties

		public NBiographicDataSchema BiographicData { get; set; }
		public NBiographicDataSchema CustomData { get; set; }
		public string GenderDataName { get; set; }
		public string EnrollDataName { get; set; }
		public string ThumbnailDataName { get; set; }
		public string SchemaName { get; set; }
		public bool HasCustomData
		{
			get
			{
				return CustomData != null && CustomData.Elements.Count > 0;
			}
		}
		public bool IsEmpty
		{
			get
			{
				return this == Empty;
			}
		}

		#endregion

		#region Public fields

		public static readonly SampleDbSchema Empty = new SampleDbSchema() { SchemaName = "None" };

		#endregion

		#region Public methods

		public override string ToString()
		{
			return SchemaName;
		}

		public static SampleDbSchema Parse(string value)
		{
			SampleDbSchema sc = new SampleDbSchema();
			var values = value.Split(new string[] { "#" }, StringSplitOptions.None);
			if (values.Length != 6) throw new ArgumentException();

			sc.SchemaName = values[0];
			sc.BiographicData = NBiographicDataSchema.Parse(values[1]);
			if (!string.IsNullOrEmpty(values[2]))
				sc.CustomData = NBiographicDataSchema.Parse(values[2]);

			sc.GenderDataName = values[3].Split(new char[] { '=' })[1];
			sc.ThumbnailDataName = values[4].Split(new char[] { '=' })[1];
			sc.EnrollDataName = values[5].Split(new char[] { '=' })[1];

			return sc;
		}

		public string Save()
		{
			if (IsEmpty) throw new InvalidOperationException();

			string format = "{0}#{1}#{2}#Gender={3}#Thumbnail={4}#EnrollData={5}";
			return string.Format(format, SchemaName, (object)BiographicData ?? string.Empty, (object)CustomData ?? string.Empty, GenderDataName, ThumbnailDataName, EnrollDataName);
		}

		#endregion
	};

	public static class SettingsManager
	{
		#region Public methods

		public static void LoadSettings(NBiometricClient client)
		{
			Settings s = Settings.Default;
			NPropertyBag propertyBag = null;
			string propertiesString = string.Empty;

			if (client == null) throw new ArgumentNullException("client");

			client.Reset();
			client.UseDeviceManager = true;
			try { propertiesString = Settings.Default.ClientProperties; }
			catch { }
			propertyBag = NPropertyBag.Parse(propertiesString);
			propertyBag.ApplyTo(client);

			client.FingersDeterminePatternClass = client.FingersDeterminePatternClass && LicensingTools.CanDetectFingerSegments(client.LocalOperations); ;
			client.FingersCalculateNfiq = client.FingersCalculateNfiq && LicensingTools.CanAssessFingerQuality(client.LocalOperations);
			var remoteConnection = client.RemoteConnections.FirstOrDefault();
			NBiometricOperations remoteOperations = remoteConnection != null ? remoteConnection.Operations : NBiometricOperations.None;
			client.FingersCheckForDuplicatesWhenCapturing = client.FingersCheckForDuplicatesWhenCapturing && LicensingTools.CanFingerBeMatched(remoteOperations);
			if (!LicensingTools.CanDetectFaceSegments(client.LocalOperations))
			{
				client.FacesDetectAllFeaturePoints = false;
				client.FacesDetectBaseFeaturePoints = false;
				client.FacesDetermineGender = false;
				client.FacesRecognizeEmotion = false;
				client.FacesDetectProperties = false;
				client.FacesRecognizeEmotion = false;
				client.FacesDetermineAge = false;
			}
		}

		public static void LoadPreferedDevices(NBiometricClient client)
		{
			Settings s = Settings.Default;
			try
			{
				if (!string.IsNullOrEmpty(s.FingerScanner))
				{
					NDevice device = client.DeviceManager.Devices.FirstOrDefault(x => x.Id == s.FingerScanner);
					if (device != null) client.FingerScanner = (NFScanner)device;
				}
			}
			catch { }
			try
			{
				if (!string.IsNullOrEmpty(s.PalmScanner))
				{
					NDevice device = client.DeviceManager.Devices.FirstOrDefault(x => x.Id == s.PalmScanner);
					if (device != null) client.PalmScanner = (NFScanner)device;
				}
			}
			catch { }
			try
			{
				if (!string.IsNullOrEmpty(s.IrisScanner))
				{
					NDevice device = client.DeviceManager.Devices.FirstOrDefault(x => x.Id == s.IrisScanner);
					if (device != null) client.IrisScanner = (NIrisScanner)device;
				}
			}
			catch { }
			try
			{
				if (!string.IsNullOrEmpty(s.FaceCaptureDevice))
				{
					NDevice device = client.DeviceManager.Devices.FirstOrDefault(x => x.Id == s.FaceCaptureDevice);
					if (device != null) client.FaceCaptureDevice = (NCamera)device;
				}
			}
			catch { }
			try
			{
				if (!string.IsNullOrEmpty(s.VoiceCaptureDevice))
				{
					NDevice device = client.DeviceManager.Devices.FirstOrDefault(x => x.Id == s.VoiceCaptureDevice);
					if (device != null) client.VoiceCaptureDevice = (NMicrophone)device;
				}
			}
			catch { }
		}

		public static void SaveSettings(NBiometricClient client)
		{
			Settings s = Settings.Default;
			NPropertyBag properties = new NPropertyBag();

			if (client == null) throw new ArgumentNullException("client");

			client.CaptureProperties(properties);
			s.ClientProperties = properties.ToString();

			// prefered devices
			s.FaceCaptureDevice = client.FaceCaptureDevice != null ? client.FaceCaptureDevice.Id : null;
			s.FingerScanner = client.FingerScanner != null ? client.FingerScanner.Id : null;
			s.PalmScanner = client.PalmScanner != null ? client.PalmScanner.Id : null;
			s.IrisScanner = client.IrisScanner != null ? client.IrisScanner.Id : null;
			s.VoiceCaptureDevice = client.VoiceCaptureDevice != null ? client.VoiceCaptureDevice.Id : null;

			s.Save();
		}

		#endregion

		#region Public properties

		public static IEnumerable<Phrase> Phrases
		{
			get
			{
				string values = null;
				try { values = Settings.Default.Phrases; }
				catch { };

				if (!string.IsNullOrEmpty(values))
				{
					string[] split = values.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);
					foreach (var item in split)
					{
						Phrase phrase = null;
						try
						{
							var splitPhrase = item.Split('=');
							phrase = new Phrase(int.Parse(splitPhrase[0]), splitPhrase[1]);
						}
						catch
						{
							//ignore invalid entries
							continue;
						};
						yield return phrase;
					}
				}
			}
			set
			{
				if (value != null)
				{
					StringBuilder phrases = new StringBuilder();
					foreach (Phrase phrase in value)
					{
						phrases.AppendFormat("{0}={1};", phrase.Id, phrase.String);
					}
					Settings.Default.Phrases = phrases.ToString();
				}
				else
				{
					Settings.Default.Phrases = string.Empty;
				}
				Settings.Default.Save();
			}
		}

		public static ConnectionType ConnectionType
		{
			get
			{
				try { return (ConnectionType)Settings.Default.ConnectionType; }
				catch { return ConnectionType.SQLiteDatabase; }
			}
			set
			{
				Settings.Default.ConnectionType = (int)value;
				Settings.Default.Save();
			}
		}

		public static string OdbcConnectionString
		{
			get
			{
				try { return Settings.Default.OdbcConnectionString; }
				catch { return string.Empty; }
			}
			set
			{
				Settings.Default.OdbcConnectionString = value;
				Settings.Default.Save();
			}
		}

		public static string TableName
		{
			get
			{
				try { return Settings.Default.TableName; }
				catch { return string.Empty; }
			}
			set
			{
				Settings.Default.TableName = value;
				Settings.Default.Save();
			}
		}

		public static string RemoteServerAddress
		{
			get
			{
				try { return Settings.Default.HostName; }
				catch { return "localhost"; }
			}
			set
			{
				Settings.Default.HostName = value;
				Settings.Default.Save();
			}
		}

		public static int RemoteServerPort
		{
			get
			{
				try { return Settings.Default.ClientPort; }
				catch { return 25452; }
			}
			set
			{
				Settings.Default.ClientPort = value;
				Settings.Default.Save();
			}
		}

		public static int RemoteServerAdminPort
		{
			get
			{
				try { return Settings.Default.AdminPort; }
				catch { return 24932; }
			}
			set
			{
				Settings.Default.AdminPort = value;
				Settings.Default.Save();
			}
		}

		public static int FingersGeneralizationRecordCount
		{
			get
			{
				try { return Settings.Default.FingersGeneralizationRecordCount; }
				catch { return 3; }
			}
			set
			{
				Settings.Default.FingersGeneralizationRecordCount = value;
				Settings.Default.Save();
			}
		}

		public static int PalmsGeneralizationRecordCount
		{
			get
			{
				try { return Settings.Default.PalmsGeneralizationRecordCount; }
				catch { return 3; }
			}
			set
			{
				Settings.Default.PalmsGeneralizationRecordCount = value;
				Settings.Default.Save();
			}
		}

		public static int FacesGeneralizationRecordCount
		{
			get
			{
				try { return Settings.Default.FacesGeneralizationRecordCount; }
				catch { return 3; }
			}
			set
			{
				Settings.Default.FacesGeneralizationRecordCount = value;
				Settings.Default.Save();
			}
		}

		public static string[] QuerySuggestions
		{
			get
			{
				try { return Settings.Default.QueryAutoComplete.OfType<string>().ToArray(); }
				catch { return new string[0]; }
			}
			set
			{
				Settings.Default.QueryAutoComplete = new System.Collections.Specialized.StringCollection();
				if (value != null)
				{
					Settings.Default.QueryAutoComplete.AddRange(value);
				}
				Settings.Default.Save();
			}
		}

		public static bool WarnHasSchema
		{
			get
			{
				try { return Settings.Default.WarnHasSchema; }
				catch { return true; }
			}
			set
			{
				Settings.Default.WarnHasSchema = value;
				Settings.Default.Save();
			}
		}

		public static int CurrentSchemaIndex
		{
			get
			{
				try { return Settings.Default.CurrentScema; }
				catch { return 0; }
			}
			set
			{
				Settings.Default.CurrentScema = value;
				Settings.Default.Save();
			}
		}

		public static IEnumerable<SampleDbSchema> Schemas
		{
			get
			{
				string[] schemas = null;
				try { schemas = Settings.Default.Schemas.OfType<string>().ToArray(); }
				catch { schemas = new string[0]; }

				return schemas.Select(x => SampleDbSchema.Parse(x)).ToArray();
			}
			set
			{
				Settings.Default.Schemas.Clear();
				if (value != null)
				{
					foreach (var item in value)
					{
						Settings.Default.Schemas.Add(item.Save());
					}
				}
				Settings.Default.Save();
			}
		}

		public static SampleDbSchema CurrentSchema
		{
			get
			{
				if (CurrentSchemaIndex == -1)
					return SampleDbSchema.Empty;
				else
					return Schemas.ToArray()[CurrentSchemaIndex];
			}
		}

		public static int LocalOperationsIndex
		{
			get
			{
				try { return Settings.Default.LocalOperations; }
				catch { return 5; }
			}
			set
			{
				Settings.Default.LocalOperations = value;
				Settings.Default.Save();
			}
		}

		public static bool FacesMirrorHorizontally
		{
			get
			{
				try { return Settings.Default.FacesMirrorHorizontally; }
				catch { return false; }
			}
			set
			{
				Settings.Default.FacesMirrorHorizontally = value;
				Settings.Default.Save();
			}
		}

		#endregion
	}
}
